### DISCUSSION TITLE
请问CommunityToolkit.Mvvm中如何绑定Grid的MouseMove事件？

### DISCUSSION THREAD

**Discussion by WWMMFFLLRR** [BYJRK/DotNet-Discussions#23] [MVVM]
![image](https://user-images.githubusercontent.com/37059082/217736411-3e15ec8c-7cba-4afe-b5c6-ee2eaf1d3aa5.png)
我是这样写的，然后编译会报错，说这个方法不能用于生成命令属性，因为其签名与任何现有的中继命令类型都不兼容。

**Comment by Hakoyu** [comment]
因为你使用的`MouseButtonEventHandler(s,e)`本身是`Event`而不是`Command`
而且`RelayCommand`也只支持一个参数
一般来说,`MouseMove`这种事件都会在`View`处理
如果执意要在`ViewModel`中处理的话,可以使用[Behaviors](https://github.com/Microsoft/XamlBehaviorsWpf)
```xaml
xmlns:i="http://schemas.microsoft.com/xaml/behaviors"
<Grid>
  <i:Interaction.Triggers>
    <i:EventTrigger EventName="MouseMove">
      <i:InvokeCommandAction Command="{Binding MouseMoveCommand}" />
    </i:EventTrigger>
  </i:Interaction.Triggers>
</Grid>
```
```csharp
[RelayCommand]
private void MouseMove()
{
    ...
}
```

**Comment by WWMMFFLLRR** [comment]
![image](https://user-images.githubusercontent.com/37059082/217791431-81f53fbe-e429-45b1-b321-614db31a8f46.png)
![image](https://user-images.githubusercontent.com/37059082/217791569-71c8b3a0-a6f7-403a-ba77-5461b272bea0.png)
你好，请问下，我这样写，为什么传入ViewModel 的MouseMove（MouseEventArgs e）的参数e为什么一直为null呢？

**Reply by WWMMFFLLRR** [reply]
<i:Interaction.Triggers>
            <i:EventTrigger EventName="MouseMove">
                <i:InvokeCommandAction
                    Command="{Binding MouseMoveCommand}"
                    PassEventArgsToCommand="True"
                    CommandParameter="{Binding RelativeSource={RelativeSource Mode=FindAncestor, AncestorType={x:Type Grid}}, Path=MouseEventArgs}">
                </i:InvokeCommandAction>
            </i:EventTrigger>

你好，这样写好像可以成功实现这个功能了，但不知道这样写好吗？

**Reply by BYJRK** [reply]
你写了 `PassEventArgsToCommand` 就不用再写 `CommandParameter` 了

**Reply by WWMMFFLLRR** [reply]
好的，我明天试一下，谢谢了。

**Comment by BYJRK** [comment]
@Hakoyu [说的](https://github.com/BYJRK/DotNet-Discussions/discussions/23#discussioncomment-4915277)没错，你如果非要把 `MouseOver` 事件绑定到 ViewModel 中，那么你恐怕只能考虑用附加属性或者行为。**但是这样是不太对的，因为这样违背了 MVVM**，因为你如果把 `MouseEventArgs` 传递到了 VM，那么就相当于 VM 知道了 View 中的信息。

所以更为正确的方式是，你需要思考一下你为什么要注册 `MouseOver`，从而选择更符合 MVVM 模式的方法。这里举几个例子：

1. 如果你想修改按钮的触发条件，也就是不再是点击，而是鼠标放上去就响应，那么 `Button` 本身有 `ClickMode` 属性，只要修改为 `ClickMode.Hover` 按钮就会在鼠标置于上方时触发 `Click` 事件
2. 如果你想在鼠标置于按钮上方时修改按钮的样式，那么你应该考虑用 Style/Template + Trigger 的方式
3. 如果你想制作一个界面效果，比如鼠标置于上方时打开 `Button` 的 `ContextMenu`，那么你应该把这个功能放在 View 中（因为这和 ViewModel 并没有什么关系），或者把功能封装到附加属性或行为中

以上内容我近期会制作一期视频简单聊一聊~敬请期待

**Reply by WWMMFFLLRR** [reply]
好的，感谢大佬提点，这样写好像是不太符合MVVM的设计，我这边ViewModel和View写的也不太好，我再重新想想怎么实现我的功能。

我刚学习WPF不久，如果各位大佬有空，希望能多多指教一下。
我要实现的是一个按住鼠标中键然后拖动鼠标修改Image控件中图像的窗宽窗位这样，ViewModel里面的MouseMove代码如下:

[RelayCommand]
private void MouseMove(MouseEventArgs e)
{
    if (e.MiddleButton == MouseButtonState.Pressed && currentDicom != null)
            {
                if (pt_Down == new System.Windows.Point(0, 0))
                {
                    pt_Down = e.GetPosition(ImageView);
                    return;
                }
                if (pt_Down == e.GetPosition(ImageView))
                {
                    return;
                }
                System.Windows.Point ptCurrent = e.GetPosition(ImageView);
                if (Math.Abs((ptCurrent.X - pt_Down.X)) > ratio || Math.Abs((ptCurrent.Y - pt_Down.Y)) > ratio)
                {
                    pt_Down = e.GetPosition(ImageView);
                    return;
                }
                this.WindowWidth += (ptCurrent.X - pt_Down.X) * ratio;
                this.WindowCenter += (ptCurrent.Y - pt_Down.Y) * ratio;
                dcmToBitmap(this.currentDicom, WindowWidth, WindowCenter);
                Update_Image();
                pt_Down = e.GetPosition(ImageView);
            }
}

View里面现在比较干净，View里面没有WindowWidth ，WindowCenter的属性 ，可能还需要一次move更新了WindowWidth ，WindowCenter后，然后用最新的WindowWidth ，WindowCenter去刷新image的source。
不知道有好的方法实现没有，如有空还希望能多多指点一下。

**Reply by BYJRK** [reply]
你其实可以考虑把这些东西全写在 View 里面，也就是 Window.xaml.cs 中，因为这和 ViewModel 实在是关系不大。刚接触 MVVM 的时候，总会想着让 xaml.cs 里面尽量干净，但是这个思路其实并不是很正确。有的时候，强行把某些功能塞进 ViewModel 就是在折磨自己，而且什么好处也得不到。常见的例子比如：

1. 点击一个按钮，让当前窗口最大化、置顶或者关闭
2. 给 TextBox 实现一个拖入外部文本并加载的功能
3. 给 Button 添加 MouseMove 事件并实时捕获鼠标的位置
4. 其他待补充……

这些也不能说是完全不能在 VM 里面实现，但是折腾了各种自定义控件、附加属性、行为和触发器之后，你会发现这一点也不优雅。所以我的观点是，如果有一些功能它就是界面的功能（比如不需要和后台的数据进行交互等），那么就把它放在 View 中实现即可

**Comment by Hakoyu** [comment]
`ViewModel`原则上是禁止任何`View`的内容
例如**WPF**,`ViewModel`中不允许使用任何`System.Windows`下的内容
也就是说不需要**WPF**,哪怕是**控制台项目中**`ViewModel`和`Model`都是可以正常运行的

**Reply by WWMMFFLLRR** [reply]
好的，确实是这样的，不能一昧追求View的干净，多谢大佬抽空指教。
